pwndbg/commands/cymbol.py

```diff
@@ -0,0 +1,253 @@
+)
+
+# Remeber loaded symbols. This would be useful for 'remove-symbol-file'.
+loaded_symbols = {}
+
+
+def unload_loaded_symbol(custom_structure_name):
+    custom_structure_symbols_file = loaded_symbols.get(custom_structure_name)
+    if custom_structure_symbols_file is not None:
+        gdb.execute(f"remove-symbol-file {custom_structure_symbols_file}")
+        loaded_symbols.pop(custom_structure_name)
+
+
+def OnlyWhenStructureExists(func):
+    @functools.wraps(func)
+    def wrapper(custom_structure_name):
+        pwndbg_cachedir = pwndbg.lib.tempfile.cachedir("custom-symbols")
+        pwndbg_custom_structure_path = os.path.join(pwndbg_cachedir, custom_structure_name) + ".c"
+        if not os.path.exists(pwndbg_custom_structure_path):
+            print(message.error("No custom structure was found with the given name!"))
+            return
+        return func(custom_structure_name, pwndbg_custom_structure_path)
+
+    return wrapper
+
+
+def generate_debug_symbols(custom_structure_path, pwndbg_debug_symbols_output_file=None):
+    if not pwndbg_debug_symbols_output_file:
+        _, pwndbg_debug_symbols_output_file = tempfile.mkstemp(prefix="custom-", suffix=".dbg")
+
+    # -fno-eliminate-unused-debug-types is a handy gcc flag that lets us extract debug symbols from non-used defined structures.
+    gcc_extra_flags = [
+        custom_structure_path,
+        "-c",
+        "-g",
+        "-fno-eliminate-unused-debug-types",
+        "-o",
+        pwndbg_debug_symbols_output_file,
+    ]
+
+    # TODO: implement remote debugging support.
+    gcc_flags = pwndbg.lib.gcc.which(pwndbg.gdblib.arch)
+    if gcc_compiler_path != "":
+        gcc_flags[0] = gcc_compiler_path
+
+    gcc_cmd = gcc_flags + gcc_extra_flags
+
+    try:
+        subprocess.run(gcc_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+    except subprocess.CalledProcessError as exception:
+        print(message.error(exception))
+        print(message.error("Parsing failed, try to fix any syntax errors first."))
+        return None
+    except Exception as exception:
+        print(message.error(exception))
+        print(message.error("An error occured while generating the debug symbols."))
+        return None
+
+    return pwndbg_debug_symbols_output_file
+
+
+def add_custom_structure(custom_structure_name):
+    pwndbg_cachedir = pwndbg.lib.tempfile.cachedir("custom-symbols")
+    pwndbg_custom_structure_path = os.path.join(pwndbg_cachedir, custom_structure_name) + ".c"
+
+    if os.path.exists(pwndbg_custom_structure_path):
+        option = input(
+            message.notice(
+                "A custom structure was found with the given name, would you like to overwrite it? [y/n] "
+            )
+        )
+        if option != "y":
+            return
+
+    print(
+        message.notice("Enter your custom structure in a C header style, press Ctrl+D to save:\n")
```

Comment:
Potential improvement: create a struct_template.c file in pwndbg and then a GDB paramater which will point to it and  whenever the user adds a struct, copy the template first and open the editor on the copied template.

Example template:
```c
// This is a Pwndbg cymbol command C structure template file
// To use a different template for cymbol C structs, use `set cymbol-template-path <path-to-c-file>` command
#include <stdio.h>
#include <stdint.h>

typedef struct  {
      // Put your structure fields here
} <custom_structure_name>;
```

Reply:
Its fine to leave this for now and add it in a future PR.